### PR TITLE
Add spectrum support

### PR DESCRIPTION
**Pull request by mtami**
Redshift Spectrum allows you to create external tables that scan objects on S3.
This PR allow you to infer data types from the table.
Also, it supports creating external schema/table.

This is a complete version of this old [PR](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/163)
I pick up where the old PR finish and add more functions.
Big thanks to @briandailey  

## Todos
- [x] MIT compatible
- [x] Tests
- [x] Documentation
- [x] Updated CHANGES.rst

### CODE REVIEW

**Comment by mtami** [comment]
Hi @jklukas ,
can u please check this PR.

**Comment by jklukas** [comment]
@Brooke-white - Would you be able to give this a look? Do you think there's potential for this slowing down reflection? I don't know what to expect from performance of lookups in `svv_columns_external`.

**Comment by Brooke-white** [comment]
Hey @mtami -- thank you for posting this PR -- I will post my review by EOD tomorrow :)

**Review by Brooke-white — CHANGES_REQUESTED** [review]
just a couple small comments, I still need to kick off a test run.

**Comment by Brooke-white** [comment]
> Do you think there's potential for this slowing down reflection? I don't know what to expect from performance of lookups in `svv_columns_external`.

@jklukas, I found an interesting discussion about performance in [this issue and linked PR from the dbt-core repo](https://github.com/dbt-labs/dbt-core/issues/2854) relating to `svv_external_columns`.

What I got out of that was that performance will degrade if the user has a ton of external tables, and there is no `table` or `schema` filter applied.

This PR applies those filters as recommended(e.g. [here](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/243/files?authenticity_token=uOqWLH4KXbougKYHOyIK44%2FuEhkzncYbhZQABg2XyY3rvVt15PWTbPUOiZzrzp1jETznhgdWEOQZCQUscEQDuw%3D%3D&file-filters%5B%5D=.py#diff-da4afac25892e729bdeca4533caf15820b269c54b5ad8bc7dbda7c47cdd76753R909-R916)).

My one thought would be the impact on existing users who fall into the bucket of having many external tables, and the potential degradation in performance they'd see if calling a method such as [`get_pk_contraint()`](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/243/files#diff-da4afac25892e729bdeca4533caf15820b269c54b5ad8bc7dbda7c47cdd76753R670) without passing anything to `schema`. Though passing something to `schema` would fix this :). 

With that said, I'm not very knowledgeable of this part of the codebase so I'm not sure how much of an issue this could be/if it would be one in the first place. I think this PR is taking the correct considerations, and having the support for external tables would be a big benefit for users :).

**Comment by jklukas** [comment]
> My one thought would be the impact on existing users who fall into the bucket of having many external tables, and the potential degradation in performance they'd see if calling a method such as [`get_pk_contraint()`](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/243/files#diff-da4afac25892e729bdeca4533caf15820b269c54b5ad8bc7dbda7c47cdd76753R670) without passing anything to `schema`. Though passing something to `schema` would fix this :).
> 
> With that said, I'm not very knowledgeable of this part of the codebase so I'm not sure how much of an issue this could be/if it would be one in the first place. I think this PR is taking the correct considerations, and having the support for external tables would be a big benefit for users :).

Thanks for doing this research! I agree with your analysis. Let's just make sure we're clear in the release notes about the change and potential performance impact and perhaps link to this discussion for how to mitigate if users do see a perf regression on upgrade.

File: CHANGES.rst [hunk 1]
```diff
@@ -1,3 +1,10 @@
+0.8.11 (unreleased)
+------------------
+
+- Add support for reflection of spectrum tables.
+  (`Issue #122 <https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/issues/122>`_)
```
[the remark this answers is not in the document]
Comment (mtami) [reply CHANGES.rst]: done

File: sqlalchemy_redshift/dialect.py [hunk 2]
```diff
@@ -860,77 +963,127 @@ def _get_all_relation_info(self, connection, **kw):
...
+        table_clause = (
+            "AND table_name = '{table}'".format(table=table_name) if table_name else ""
         )
-        for col in result:
-            key = RelationKey(col.table_name, col.schema, connection)
-            all_columns[key].append(col)
+
+        all_columns = defaultdict(list)
+        with connection.connect() as cc:
+            result = cc.execute(sa.text("""
```
[the remark this answers is not in the document]
Comment (mtami) [reply sqlalchemy_redshift/dialect.py:979]: Hey @Brooke-white,
Thank you for pointing this.
I update the external types that require alias mapping.
For reference, what i do is creating two identical tables (external & internal); then i do the map.

>  CASE
>          WHEN c.external_type = 'int' THEN 'integer'
>          WHEN c.external_type = 'float' THEN 'real'
>          WHEN c.external_type = 'double' THEN 'double precision'
>          WHEN c.external_type = 'timestamp' THEN 'timestamp without time zone'
>          ELSE
>            replace(
>             replace(
>                 replace(c.external_type, 'decimal', 'numeric'),
>                 'char', 'character'),
>             'varchar', 'character varying')
>    END
>

**Comment by Brooke-white** [comment]
@mtami , I kicked off CI but saw some failures. Could you please take a look?

**Comment by mtami** [comment]
> @mtami , I kicked off CI but saw some failures. Could you please take a look?

hey @Brooke-white, 
The issue [raised here](https://app.travis-ci.com/github/sqlalchemy-redshift/sqlalchemy-redshift/jobs/553293108#L353) is from `test_pk_name` method under test path.

I successfully can run the following snippet locally in both Py [3.9.0, 3.5.3]:

```
redshift_session = sa.orm.Session(engine)
dialect = redshift_session.bind.dialect
pk = dialect.get_pk_constraint(
  redshift_session.connection(),
  "my_table"
)
print(pk)


```

However, when i edit the above snippet slightly to use `with` statement, I got an error.
```
with sa.orm.Session(engine) as redshift_session: # <- AttributeError: __enter__
    ...

```
The ci logs show similar [error message](https://app.travis-ci.com/github/sqlalchemy-redshift/sqlalchemy-redshift/jobs/553293108#L385-L390).
Unfortunately, I am not sure what cause this error or how to solve it.

But it seems to me, this error has nothing to do with current PR (maybe?), it might be something related to how sqlalchemy-redshift expose redshift session internally to pytest.

**Comment by Brooke-white** [comment]
CI is passing on `main` branch, and while there are some slight differences (mainly a different Redshift cluster), my local run shows `main` passing, and the same error when using the PR branch. 

From the debugger...so far what I'm seeing is that this exception is happening when applying database migrations in `conftest.py/migrated_database()`. the call to `self.migrate(engine)` raises an error e.g. `{NoSuchTableError}public.reflection_default_value`, which leads to `with _redshift_database_tool.migrated_database() as egs:` in `contest.py/_session_scoped_redshift_engine(_redshift_database_tool)` failing as the return value is not as expected given the use of `with` here. [ref](https://stackoverflow.com/questions/61715302/i-get-the-following-error-attributeerror-enter-when-only-using-one-with-blo)

When looking at this issue I'm just debugging one test, `tests/test_constraint_names.py`, using `tox`, since the issue is repeated, and just looking at this one test makes things easier for debugging.

To do this I make the following modifications to run the CI tests with my own cluster, this allows me to use the PyCharm debugger with tox which is really helpful for not so straight forward issues like this. What is done below comments out a bunch of logic used for bringing up the CI Redshift cluster and just hardcodes in the credentials for my testing cluster. Also, I modify `tox.ini` to run a single test file with a single configuration (py35-sa13) and just one DB-API (psycopg2).

```diff
diff --git a/tests/conftest.py b/tests/conftest.py
index 69ae1a1..c92427b 100644
--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -138,24 +138,24 @@ def pytest_generate_tests(metafunc):
 @pytest.yield_fixture(scope='session')
 def _redshift_database_tool(redshift_dialect_flavor):
     from rs_sqla_test_utils import models
-    if 'PGPASSWORD' not in os.environ:
-        pytest.skip('This test will only work on Travis.')
-
-    session = requests.Session()
-
-    while True:
-        resp = session.post('https://bigcrunch.herokuapp.com/session/')
-        if resp.status_code != 503:
-            break
-        print('waiting for Redshift to boot up')
-        time.sleep(15)
-
-    resp.raise_for_status()
-    config = resp.json()
+    # if 'PGPASSWORD' not in os.environ:
+    #     pytest.skip('This test will only work on Travis.')
+    #
+    # session = requests.Session()
+    #
+    # while True:
+    #     resp = session.post('https://bigcrunch.herokuapp.com/session/')
+    #     if resp.status_code != 503:
+    #         break
+    #     print('waiting for Redshift to boot up')
+    #     time.sleep(15)
+    #
+    # resp.raise_for_status()
+    # config = resp.json()
     try:
         class RedshiftDatabaseTool(DatabaseTool):
             engine_definition = db.redshift_engine_definition(
-                config['cluster'],
+                None,#config['cluster'],
                 redshift_dialect_flavor
             )
 
@@ -164,11 +164,12 @@ def _redshift_database_tool(redshift_dialect_flavor):
 
         yield RedshiftDatabaseTool()
     finally:
-        base_url = resp.request.url
-        resp = session.delete(
-            urlparse.urljoin(base_url, config['resource_url'])
-        )
-        resp.raise_for_status()
+        pass
+        # base_url = resp.request.url
+        # resp = session.delete(
+        #     urlparse.urljoin(base_url, config['resource_url'])
+        # )
+        # resp.raise_for_status()
 
 
 @pytest.yield_fixture(scope='function')
diff --git a/tests/rs_sqla_test_utils/db.py b/tests/rs_sqla_test_utils/db.py
index 5f73206..1552408 100644
--- a/tests/rs_sqla_test_utils/db.py
+++ b/tests/rs_sqla_test_utils/db.py
@@ -26,10 +26,10 @@ def redshift_engine_definition(cluster, dialect):
     return EngineDefinition(
         db_connect_url=sa_url.URL(
             drivername=dialect,
-            username='travis',
-            password=os.environ['PGPASSWORD'],
-            host=cluster['Address'],
-            port=cluster['Port'],
+            username='awsuser',
+            password='xxx',
+            host='xxx,
+            port=5439,
             database='dev',
             query={'client_encoding': 'utf8'},
         ),
diff --git a/tox.ini b/tox.ini
index bb73c55..3c0b531 100644
--- a/tox.ini
+++ b/tox.ini
@@ -1,9 +1,9 @@
 [tox]
 envlist =
     py35-sa13,
-    py39-sa14,
-    lint,
-    docs
+;    py39-sa14,
+;    lint,
+;    docs
 
 [testenv]
 passenv = PGPASSWORD
@@ -12,7 +12,7 @@ deps =
     sa14: sqlalchemy==1.4.15
 
 [testenv:py35-sa13]
-commands = pytest -s {posargs} --dbdriver psycopg2 --dbdriver psycopg2cffi
+commands = pytest -s tests/test_constraint_names.py --dbdriver psycopg2
 deps =
     alembic==1.4.2
     packaging==20.4
```

I'll be heading out for the holidays shortly, but will update here if I find anything else before then. I'll pick this back up in the new year. Anyways, happy holidays and best of luck!

**Comment by mtami** [comment]
Hi @Brooke-white ,
can you kick off CI again, I think the latest commit will solve the issue.

**Comment by Brooke-white** [comment]
@mtami -- I've kicked off another run and am seeing the earlier failures are resolved, though there is a new failure seen `database "testdb_xxx" is being accessed by other users`

**Comment by mtami** [comment]
@Brooke-white - This error shown when teardown of **test_external_table_reflection** because this specific test is executed in _AUTOCOMMIT_  level rather that _TRANSACTION_  (with concurrent run PY39, PY35), trying to figure a work around  to run this test in concurrent way, any idea ?

**Comment by Brooke-white** [comment]
Thanks for those details, @mtami.

I'm taking a crack at reproducing this locally using tox's [`--parallel`](https://tox.wiki/en/latest/example/basic.html#parallel-mode) to try and mirror how tests run on Travis.

I'm thinking adding some "wait" logic to the test, to ensure there is no one else accessing the db could be worthwhile to investigate, though there are multiple statements that need to be executed sequentially...I will update shortly with any findings.

**Comment by svdimchenko** [comment]
any update about this PR ?

**Comment by Brooke-white** [comment]
Hey @svdimchenko , apologies for the delays. From what I remember, the failure seen had to do with `test_external_table_reflection()` needing to execute statements in parallel (create, query, drop the external schema), and that being incompatible with how tests are run in travis (in parallel), giving this other users are accessing the db error.

I also want to mention [this](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/163#issuecomment-479191247) conversation chain re: IAM role needed for testing. @jklukas mentioned he created the necessary role, but given it was a while ago, let's make sure it still exists. Otherwise we will run into this issue after the current one is resolved :).

I don't have the time to troubleshoot the issue with the tests right now, but I'm more than happy to provide any support to @mtami or others who might have the bandwidth to look into this a bit more :). I believe `tox --parallel` helped repro this locally.

**Comment by mdeshmu** [comment]
any update about this PR ?

**Comment by Brooke-white** [comment]
This PR is blocked by the above mentioned issue with the tests. I don't have the bandwidth to look into the failures more at this time, but happy to help out where I can if someone else is able to dig into this more :)

**Comment by mdeshmu** [comment]
@mtami can you please look into this further if you have bandwidth.

**Comment by mtami** [comment]
Hi @Brooke-white,
Currently I don't have access to redshift cluster.
If you can help in providing access for a redshift cluster for testing purposes.

**Comment by WTFox** [comment]
Hello! We're doing some upgrades and would love to see work continue on this PR. I'm happy to help and pick up where work left off. I have a few questions on how to best approach this. Would you prefer I pick up this PR start where it leaves off or start fresh from main? Do you have any documentation on contributing? Would it be worth kicking off a fresh CI run to see where things stand or should we try to rebase against main first? [This comment](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/243#issuecomment-1027109938) is extremely helpful!

**Comment by Brooke-white** [comment]
Hi @WTFox , feel free to work off this PR. For contributing please review our Code of Conduct and PR template for what's required. You can let me know if you have any specific questions not answered by those resources.

I'll kick off a run of our integration test suite for what's here now, and then I'd recommend doing a rebase/merge from main.

**Comment by WTFox** [comment]
@Brooke-white I believe we can close this since these commits were brought in when #263 was merged.

**Pull request closed without merging**